sdk/lib/queue/queue.cc

```diff
@@ -0,0 +1,437 @@
+#include <locks.hh>
+#include <queue.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <timeout.h>
+#include <type_traits>
+
+using namespace CHERI;
+using cheriot::atomic;
+
+using Debug = ConditionalDebug<true, "Queue library">;
+
+#ifdef __cplusplus
+using MessageQueueCounter = atomic<uint32_t>;
+#else
+typedef uint32_t MessageQueueCounter;
+#endif
+
+#include <assembly-helpers.h>
+
+namespace
+{
+	/**
+	 * Helpers for the queue.  The queue uses two counters that wrap on double
+	 * the number of elements.  This ensures that full and empty conditions
+	 * have different values: a queue is full when the producer is one queue
+	 * length ahead of the consumer, and empty when the producer and consumer
+	 * are equal.
+	 */
+
+	/**
+	 * Helper for wrapping increment.  Increments `counter`, wrapping to zero
+	 * if it reaches double `size`.
+	 */
+	constexpr uint32_t increment_and_wrap(uint32_t size, uint32_t counter)
+	{
+		counter++;
+		if (2 * size == counter)
+		{
+			return 0;
+		}
+		return counter;
+	}
+
+	/**
+	 * Returns true if and only if the `size`, `producer`, and `consumer`
+	 * counters indicate a full queue.
+	 */
+	constexpr bool is_full(uint32_t size, uint32_t producer, uint32_t consumer)
+	{
+		return producer - consumer == size;
+	}
+
+	/**
+	 * Returns true if and only if the `producer`, and `consumer` counters
+	 * indicate an empty queue.
+	 */
+	constexpr bool is_empty(uint32_t producer, uint32_t consumer)
+	{
+		return producer == consumer;
+	}
+
+	static_assert(is_full(2, 2, 0), "is-full calculation is incorrect");
+
+	/**
+	 * Returns a pointer to the element in the queue indicated by `counter`.
+	 */
+	Capability<void> buffer_at_counter(struct QueueHandle &handle,
+	                                   uint32_t            counter)
+	{
+		// Handle wrap for the second run around the counter.
+		size_t index =
+		  counter >= handle.queueSize ? counter - handle.queueSize : counter;
+		auto             offset = index * handle.elementSize;
+		Capability<void> pointer{handle.buffer};
+		pointer.address() += offset;
+		return pointer;
+	}
+
+	/**
+	 * Flag lock that uses the high bit of a word for the lock.
+	 *
+	 * The remaining bits are free for other uses, but should not be modified
+	 * while the lock is not held.
+	 */
+	struct HighBitFlagLock
+	{
+		/**
+		 * A reference to the word that holds the lock.
+		 */
+		atomic<uint32_t> &lockWord;
+
+		/**
+		 * The bit to use for the lock.
+		 */
+		static constexpr uint32_t LockBit    = 1U << 31;
+		static constexpr uint32_t WaitersBit = 1U << 30;
+
+		// Function required to conform to the Lock concept.
+		void lock()
+		{
+			__builtin_unreachable();
+		}
+
+		static constexpr uint32_t reserved_bits()
+		{
+			return LockBit | WaitersBit;
+		}
+
+		/**
+		 * Try to acquire the lock.  Returns true on success, false on failure.
+		 */
+		bool try_lock(Timeout *t)
+		{
+			uint32_t value;
+			do
+			{
+				value = lockWord.load();
+				Debug::log("Lock word: {}", value);
```

Comment:
Not sure about the merits of keeping these `Debug::log`s around, but if we do, they should probably also include `&lockWord`, not just `value`.